// @explorelearning/utils/src/index.ts
// import cancellable from '@explorelearning/cancellable';
// import capitalize from './functions/capitalize';
// import copy from './functions/copy';
// import debounce from './functions/debounce';
// import setIfThere from './functions/setIfThere';
// import valueFromPath from './functions/valueFromPath';
// import mapByProp from './functions/mapByProp';
import { versionTag } from '@explorelearning/logger';
import packageJson from '../package.json';
versionTag(packageJson);

export { default as cancellable } from '@explorelearning/cancellable';
export { default as capitalize } from './functions/capitalize';
export { default as copy } from './functions/copy';
export { default as debounce } from './functions/debounce';
export { default as setIfThere } from './functions/setIfThere';
export { default as valueFromPath } from './functions/valueFromPath';
export { default as mapByProp } from './functions/mapByProp';

// export default {
//   cancellable,
//   capitalize,
//   copy,
//   debounce,
//   setIfThere,
//   valueFromPath,
//   mapByProp
// };
